dhall/src/Dhall.hs

```diff
@@ -900,6 +909,103 @@ instance (Interpret a, Interpret b) => Interpret (a, b)
 auto :: Interpret a => Type a
 auto = autoWith defaultInterpretOptions
 
+newtype DhallFix f t = DhallFix { unDhallFix :: t }
+newtype FixVar = FixVar { unFixVar :: (Expr Src X) }
+
+-- This is a helper instance used from implementation of Interpret for DhallFix
+instance Interpret FixVar where
+    autoWith _ = Type {..}
+        where
+        extract (App _ e0) = pure (FixVar e0)
+        extract e          = typeError expected e
+
+        expected = "t"
+
+type InterpretFix f t =
+   ( Traversable f
+   , Corecursive t
+   , Interpret (f FixVar)
+   , Base t ~ f
+   )
+
+instance InterpretFix f t => Interpret (DhallFix f t) where
+    autoWith opts = Type { extract = extr, expected = expe }
+        where
+        extr (Lam _ _ (Lam _ _ e)) = DhallFix <$> buildF e
+        extr e                     = typeError expe e
+
+        -- buildF = elgot f g <=< extract (auto @FixVar)
+        viaMonadic a b x = fromMonadic $ toMonadic . a =<< toMonadic (b x)
+        buildF = elgot f g `viaMonadic` extract (auto :: Type FixVar)
+
+        g = repack . extract xF_t . unFixVar
+        repack (Failure e) = Left (Failure e)
+        repack (Success s) = Right (Free s)
+        f (Free s) = Data.Functor.Foldable.embed <$> sequenceA s
+        f (Pure x) = typeError expe x
+
+        expe = Pi "t" (Dhall.Core.Const Dhall.Core.Type) -- = ∀(t : Type)
+             $ Pi "_" (Pi "_" (expected xF_t) "t")       -- → ∀(F t → t)
+             $ "t"                                       -- → t
+
+        xF_t = autoWith opts :: Type (f FixVar)
+
+{-| `autoWithFix` allows you to implement `Interpret` instance for a recursive
+    data type. For example, given a Haskell expression:
+
+  > data Expr
+  >    = Lit Natural
+  >    | Add Expr Expr
+  >    | Mul Expr Expr
+  >    deriving (Generic, Show)
+  >
+  > makeBaseFunctor ''Expr
+  > deriving instance Generic (ExprF a)
+  > instance Interpret a => Interpret (ExprF a)
+  > instance Interpret Expr where autoWith = autoWithFix
+
+    And a dhall file:
+
+  > -- expr.dhall
+  >
+  > -- F-Algebra declaration.
+  > let ExprF
+  >   = λ(t : Type) →
+  >   < LitF : { _1 : Natural }
+  >   | AddF : { _1 : t, _2 : t }
+  >   | MulF : { _1 : t, _2 : t }
+  >   >
+  >
+  > -- Helper function.
+  > let Expr
+  >   = λ(t : Type)
+  >   → λ(fix : ExprF t → t)
+  >   → let E = ExprF t in
+  >     { Lit = λ(x : Natural)      → fix (E.LitF { _1 = x })
+  >     , Add = λ(a : t) → λ(b : t) → fix (E.AddF { _1 = a, _2 = b })
+  >     , Mul = λ(a : t) → λ(b : t) → fix (E.MulF { _1 = a, _2 = b })
+  >     }
+  >
+  > -- Example usage.
+  > let result
+  >   = λ(t : Type)
+  >   → λ(fix : ExprF t → t)
```

Comment:
I think we can simplify the example by (A) factoring out the shared bound variables (`t` and `fix`) and (B) splitting the record of utility constructors into individual `let` bindings, like this:

```dhall
  λ(t : Type)
→ let ExprF =
        < LitF :
            { _1 : Natural }
        | AddF :
            { _1 : t, _2 : t }
        | MulF :
            { _1 : t, _2 : t }
        >
  
  in    λ(fix : ExprF → t)
      → let Lit = λ(x : Natural) → fix (ExprF.LitF { _1 = x })
        
        let Add = λ(a : t) → λ(b : t) → fix (ExprF.AddF { _1 = a, _2 = b })
        
        let Mul = λ(a : t) → λ(b : t) → fix (ExprF.MulF { _1 = a, _2 = b })
        
        in  Add (Mul (Lit 3) (Lit 7)) (Add (Lit 1) (Lit 2))
```

Reply:
Yeah, I sort of wrote it this way to suggest a way of using this that's more pleasant to the user. We could then imagine loading `Expr` and `ExprF` from a file or even adding it to the context from Haskell, and then the actual use site is reduced to simply:

```dhall
    λ(t : Type)
  → λ(fix : ExprF t → t)
  → let { Lit, Add, Mul } = Expr t fix -- Once this syntax lands.
    in -- <some expression>
```

But I guess that could be a bit much for a use example in documentation and it's better to keep it succinct. Anyway, that's my justification but I can go with either, so it's your call.

Reply:
How about leading with the fully-inline example, and then a follow-up example showing how it could be split into reusable files like you suggest?

The main thing I'm going for is to give the reader a simpler starting point that they can use to reason about the more sophisticated case.  Working through the inlined representation was what helped me understand the feature better